query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -57,96 +47,33 @@
-      case ATTRIBUTE_EXPRESSION:
-        return this.transformAttributeExpression(expression.getAttributeExpression());
-      case FUNCTION:
-        return this.transformFunction(expression.getFunction())
-            .map(expression.toBuilder()::setFunction)
-            .map(Expression.Builder::build);
-      case ORDERBY:
-        return this.transformOrderBy(expression.getOrderBy())
-            .map(expression.toBuilder()::setOrderBy)
-            .map(Expression.Builder::build);
-      case LITERAL:
-      case VALUE_NOT_SET:
-      default:
-        return Single.just(expression);
-    }
-  }
-
-  private Single<Expression> transformColumnIdentifier(ColumnIdentifier columnIdentifier) {
-    return this.projectAttributeIfPossible(columnIdentifier.getColumnName())
-        .map(expression -> this.aliasToMatchOriginal(getOriginalKey(columnIdentifier), expression))
-        .defaultIfEmpty(Expression.newBuilder().setColumnIdentifier(columnIdentifier).build());
-  }
-
-  private Single<Expression> transformAttributeExpression(AttributeExpression attributeExpression) {
+  @Override
+  protected Single<Expression> transformAttributeExpression(
+      AttributeExpression attributeExpression) {
     return this.projectAttributeIfPossible(attributeExpression.getAttributeId())
         .map(
-            expression ->
-                this.aliasToMatchOriginal(getOriginalKey(attributeExpression), expression))
+            projectedExpression ->
+                attributeExpression.hasSubpath()
+                    ? this.addSubpathOrThrow(projectedExpression, attributeExpression.getSubpath())
+                    : projectedExpression)
+        .map(expression -> this.aliasToMatchOriginal(attributeExpression, expression))
         .defaultIfEmpty(
             Expression.newBuilder().setAttributeExpression(attributeExpression).build());
   }
 
-  private Single<Function> transformFunction(Function function) {
-    return this.transformExpressionList(function.getArgumentsList())
-        .map(expressions -> function.toBuilder().clearArguments().addAllArguments(expressions))
-        .map(Function.Builder::build);
-  }
-
-  private Single<List<OrderByExpression>> transformOrderByList(
-      List<OrderByExpression> orderByList) {
-    return Observable.fromIterable(orderByList).concatMapSingle(this::transformOrderBy).toList();
-  }
-
-  private Single<OrderByExpression> transformOrderBy(OrderByExpression orderBy) {
-    return this.transformExpression(orderBy.getExpression())
-        .map(orderBy.toBuilder()::setExpression)
-        .map(OrderByExpression.Builder::build);
-  }
-
-  private Single<Filter> transformFilter(Filter filter) {
-    if (filter.equals(Filter.getDefaultInstance())) {
-      return Single.just(filter);
+  private Expression addSubpathOrThrow(Expression projectedExpression, String subpath) {
+    if (!QueryRequestUtil.isSimpleAttributeExpression(projectedExpression)) {
```

Comment:
Can you give one example or add a test for this when we hit this condition?

A projected attribute is already having subpath?

Reply:
Yeah, will add a test. As a quick example, imagine if we rename a column and so use a simple projection to point one attribute name at another. In this hypothetical, an old name of `EVENT.spanTags` becomes `EVENT.attributes`. An arriving attribute expression may still be using the old name, so when we project
```json
       "attributeExpression": {
          "attributeId": "EVENT.spanTags",
          "subpath": "span.kind"
        }
```

the projection is only based on the attribute itself, so above we first project:
```json
       "attributeExpression": {
          "attributeId": "EVENT.attributes"
        }
```
then this function is meant to restore the subpath part giving us our final projection:
```json
       "attributeExpression": {
          "attributeId": "EVENT.attributes",
          "subpath": "span.kind"
        }
```

It currently throws if the projection did not result in another simple attribute expression, since if the result was already a complex expression (either a function or map selection), it would be more difficult/potentially impossible to support (to be honest, I didn't go too far down that path - we don't have that use case at the moment and can cross that bridge when/if we get there).

Reply:
Test added in #132